### ISSUE TITLE
Deployment error due to .Net version 3.1.301 SDK on Azure Web App

### ISSUE THREAD

**Issue by timoleo23** [projectkudu/kudu#3177]
Since the default .Net SDK on Azure Web App has moved to 3.1.301, some .Net app deployments (deploy.cmd) fails due to "Access denied"

The error comes from the deployment of the .Net app on the web app with the following error message
`Command: deploy.cmd Handling ASP.NET Core Web Application deployment. Restoring NuGet packages MSBUILD : error MSB1025: An internal failure occurred while running MSBuild. System.ComponentModel.Win32Exception (5): Access is denied. at System.Diagnostics.Process.set_PriorityClassCore(ProcessPriorityClass value) at System.Diagnostics.Process.set_PriorityClass(ProcessPriorityClass value) Unhandled exception. System.ComponentModel.Win32Exception (5): Access is denied. at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine) at System.Diagnostics.Process.set_PriorityClassCore(ProcessPriorityClass value) Failed exitCode=-532462766, command=dotnet restore "D:\home\site\repository\Source\Microsoft.Teams.Apps.CompanyCommunicator.sln" at System.Diagnostics.Process.set_PriorityClass(ProcessPriorityClass value) An error has occurred during web site deployment. at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine) at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args) Unhandled exception. System.ComponentModel.Win32Exception (5): Access is denied.\r\n at System.Diagnostics.Process.set_PriorityClassCore(ProcessPriorityClass value)\r\n at System.Diagnostics.Process.set_PriorityClass(ProcessPriorityClass value)\r\n at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)\r\n at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args)\r\nD:\Program Files (x86)\SiteExtensions\Kudu\87.20522.4610\bin\Scripts\starter.cmd deploy.cmd`

Here is an app template to reproduce the error - It comes form the Microsoft Communitaor App for Teams 
https://github.com/OfficeDev/microsoft-teams-company-communicator-app

Error generated from the ARM tempalte deployment 
`{ "status": "Failed", "error": { "code": "ResourceDeploymentFailure", "message": "The resource operation completed with terminal provisioning state 'Failed'." } }`

A workaround can be found based on the instructiions provided here asking for a specific version of the .Net SDK (3.1.202)
https://stackoverflow.com/questions/62595408/error-access-is-denied-deployment-to-azure-app-services

This unfortunaltely works only if the web app has the 2 verions installed (which may not be the case with new deployments)

**Comment by suwatch** [comment]
# Issue
The issue is understood and recently introduced by latest dotnet sdk (2.1.514 and 3.1.202) versions.  This impacts the projects with custom deployment script still using `dotnet restore and publish` to build.   The newly generated deployment script is now using `msbuild` and has no issue.   We are working on the fix and plan to roll out solution next week (ETA /10/2020).

# Mitigation
There is a mitigation that ones can do to unblock till the fix roll out.  Users need to update their deployment script to use msbuild as follow. 

## Fix for dotnet restore

Do change below 
```
call :ExecuteCmd dotnet restore "%DEPLOYMENT_SOURCE%\YourProject.sln"
```
to
```
call :ExecuteCmd msbuild.exe -t:Restore "%DEPLOYMENT_SOURCE%\YourProject.sln"
```

## Fix for dotnet publish

Do change below 
```
call :ExecuteCmd dotnet publish "%DEPLOYMENT_SOURCE%\YourProject.csproj" --output "%DEPLOYMENT_TEMP%" --configuration Release
```
to
```
call :ExecuteCmd msbuild.exe -t:Publish "%DEPLOYMENT_SOURCE%\YourProject.csproj" -p:OutputPath="%DEPLOYMENT_TEMP%" -p:Configuration=Release
```

That's it.

**Comment by rainersigwald** [comment]
The underlying issue is microsoft/msbuild#5365 and should be present only in 3.1.301, not 2.1.515.

**Comment by RoyalMugu** [comment]
@suwatch , I am not using any deployment script. I can repro this issue when I deploy from Visual Studio. How can I workaround this issue?

**Comment by suwatch** [comment]
@RoyalMugu  So your error is from the client side?  Meaning VS build failing before pushing artifact to App Service.  If so, mind opening another issue?

**Comment by keharvey383** [comment]
@suwatch I get the error "error MSB4057: The target "Restore" does not exist in the project. " when I try the suggested mitigation.

**Comment by suwatch** [comment]
Thanks for reporting.   Could you try follow this post https://stackoverflow.com/questions/47313537/msbuild-restore-target-msb4057-the-target-restore-does-not-exist-in-the-pro?   Or https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restoring-and-building-with-one-msbuild-command?  See if either of those addresses the issue.

**Comment by keharvey383** [comment]
@suwatch I tried "msbuild -t:build -restore" and got the error "MSBUILD : error MSB1001: Unknown switch.
Switch: -restore"

**Comment by suwatch** [comment]
are you using this in the context of deploy.cmd?   maybe you are using the old msbuild that is not support such target.   The expected msbuild should be ..

```
"D:\Program Files (x86)\MSBuild-16.4\MSBuild\Current\Bin\MSBuild.exe" -version
16.4.0.56107
```

**Comment by arjendevries** [comment]
Having the same exact issues as @keharvey383.  Tried the lasted suggestion of changing msbuild path in deploy.cmd and getting this error:  error MSB1001: Unknown switch

```SET MSBUILD_PATH=%ProgramFiles(x86)%\MSBuild-16.4\MSBuild\Current\Bin\MSBuild.exe```

**Comment by keharvey383** [comment]
@suwatch  yes I'm using it in the deploy.cmd and I did a version check and it appears to be an older version in my app service in Azure.  I'm not sure how to get it updated. 

`D:\home>msbuild.exe /version
Microsoft (R) Build Engine version 14.0.23107.0
Copyright (C) Microsoft Corporation. All rights reserved.

14.0.23107.0`

**Comment by suwatch** [comment]
Could you create another issue to discuss this?  Thanks.

**Comment by keharvey383** [comment]
@suwatch I think I have it working after looking at @arjendevries comment. 

I did the following in my deploy.cmd

SET MSBUILD_PATH=%ProgramFiles(x86)%\MSBuild-16.4\MSBuild\Current\Bin\MSBuild.exe

call :ExecuteCmd "%MSBUILD_PATH%" -t:Restore "%DEPLOYMENT_SOURCE%\my-solution.sln"

call :ExecuteCmd "%MSBUILD_PATH%" -t:Publish "%DEPLOYMENT_SOURCE%\vstar-next\my-proj.csproj" -p:OutputPath="%DEPLOYMENT_TEMP%" -p:Configuration=Dev

**Comment by WetTheSystem** [comment]
I'm trying to work through a similar issue that changing the deployment script did not resolve. I'm using the default EchoBot project from Azure and trying to run deploy in Kudu console using the changes suggested by @keharvey383 . 

I get a "error MSB4057: The target "Restore" does not exist in the project."

Link to the Microsoft thread: https://docs.microsoft.com/en-us/answers/questions/43392/how-do-i-resolve-access-denied-error-when-trying-t.html

**Comment by suwatch** [comment]
@WetTheSystem  you are using `%ProgramFiles(x86)%\MSBuild-16.4\MSBuild\Current\Bin\MSBuild.exe` and still get the error?   Please run this MSBUILD with restore explicitly on Debug Console and share the command line as well as error output.
